Aggregate saved changes for all updates inside of a transaction

bdurand opened the pull request:
### Motivation / Background

This Pull Request has been created because there is inconsistent behavior with dirty attributes in transaction callbacks when a record is updated multiple times inside of a transaction.

Consider this case:

```ruby
class User < ApplicationRecord
  after_commit :notify_email_changes, if: :email_changed?

  def notify_email_changes
    NotifyEmailChangesJob.perform_later(id)
  end
end
```

This breaks if a record is saved twice in a single transaction.

```ruby
user.transaction do
  user.update!(email: params[:email])
  if user.last_visited_at < 1.day.ago
    user.update!(last_visited_at: Time.now)
  end
end
```

In the case where we update the `last_visited_at` field, the `email_changed?` method will return false since the email address was not changed in the last save operation and `notify_email_changes` method will not be called.

Fixes #49898, #48077

### Detail

This Pull Request changes how `saved_changes` is calculated in transactions when a record is updated multiple times.

After each save operation, the list of saved changes is pushed onto a list. `saved_changes` in an `after_save` callback is unaffected and will show only the changes for that save operation.

Before a transaction is committed, the list of saved changes for all saves in the transaction are merged by reapplying them to the original attributes. Checking `saved_changes` in an `after_commit` callback will use the merged mutation tracker which includes all changes. Checking `saved_changes` outside the transaction after it is committed will also include all changes.

### Checklist

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

rafaelfranca reviewed (commented):
The idea makes sense, but the implementation isn't correct. All the `rollup_mutations_for_transaction! ` implementation is in the wrong place. Transactions should not know about the mutation tracker.

At the moment I don't have a better suggestion because I didn't investigate the problem enough, but I do know that current implementation doesn't seems correct.

That is the same problem that https://github.com/dylanahsmith/ar_transaction_changes tries to solve. It doesn't reach for the mutation tracker for example.

bdurand committed:
Aggregate saved changes for all updates inside of a transaction.

When multiple updates are made to a record inside of a transaction, each update resets the dirty fields on the record so that the `saved_changes` hash only contains the most recent change after a record is saved.

When a transaction is committed, all of the changes are aggregated into a single change set. This fixes issues where checking `saved_changes` in an `after_commit` callback could return inconsistent results depending on the order of updates in the transaction.

Fixes #49898, #48077

bdurand commented:
@rafaelfranca I would agree that it is not ideal. I've moved the bulk of the code to `active_record/attribute_methods/dirty.rb`.

I still needed to hook into the code from the transactions code since the code to roll up the saved changes needs to happen when the transaction is committed. I think this is preferable to having dirty attributes hook into transaction lifecycles.

I think using the mutation tracker is preferable to the solution in the `at_transaction_changes` gem since it provides consistency by being able to use `saved_changes` and the entire ActiveModel::Dirty API. It makes things like using `:if` filters on an `after_commit` callback work as expected (see example in the PR description). It also removes the possibility of bugs caused by developers using `saved_changes` since it work most of the time (which is the current situation).

bdurand commented:
@rafaelfranca Any thoughts on the changes made to this pull request?

pfeiffer commented:
This would be great to land! We've spent the better part of a day trying to figure out why emails were not being sent as part of our `after_commit` callbacks.

bdurand commented:
@pfeiffer FYI I did port my fix to a gem: [after_commit_changes](https://github.com/bdurand/after_commit_changes) so I could address my immediate issues while waiting for this to get picked up. This solved a number of subtle bugs in my application that I didn't even know were there.

It would be great to get this in since I'm pretty sure we're not the only ones with these issues.

jdelStrother commented:
I was about to file a bug on this but looks like it's a known problem... In case it helps any, here's the testcase I was about to post:

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails" # , github: "bdurand/rails", branch: "after_commit_dirty"
  gem "sqlite3"
end

require "active_record/railtie"
require "minitest/autorun"
ENV["DATABASE_URL"] = "sqlite3::memory:"
class TestApp < Rails::Application
  config.load_defaults Rails::VERSION::STRING.to_f
  config.eager_load = false
  config.logger = Logger.new($stdout)
  config.log_level = :warn
  config.secret_key_base = "secret_key_base"
end
Rails.application.initialize!

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :state
    t.string :foo
    t.references :channel
  end
end

class Post < ActiveRecord::Base
  after_save do
    if saved_change_to_state? && state == "active"
      update!(foo: "bar")
    end
  end

  class_attribute :state_changes
  self.state_changes = []

  # after_save do
  after_commit do
    if saved_change_to_state?
      self.state_changes << saved_change_to_state
    end
  end
end

class BugTest < ActiveSupport::TestCase
  def test_association_stuff
    post = Post.create!(state: "pending")
    post.update!(state: "active")

    assert_equal [[nil, "pending"], ["pending", "active"]], Post.state_changes
  end
end
```

which fails with after_commit not seeing the pending->active transition.

This PR fixes it. My only reservation with it is that even with the fix, `after_save` still doesn't see all of the transitions. That is, if my model looked like this - 

```ruby
class Post < ActiveRecord::Base
  after_save do
    if saved_change_to_state? && state == "active"
      update!(foo: "bar")
    end
  end

  class_attribute :state_changes
  self.state_changes = []

  after_save do
    if saved_change_to_state?
      self.state_changes << saved_change_to_state
    end
  end
end
```

that second after_save block only sees the state change from nil->pending, and not from pending->active. Which is maybe to be expected given this PR is about fixing after_commit, but still seems like unexpected behaviour.
